keps/sig-instrumentation/2580-custom-metrics-router/README.md

```diff
@@ -0,0 +1,658 @@
+using HPAs to scale based on metrics from different sources both within and
+outside the cluster.
+
+## Motivation
+
+Currently, there can only be a single [custom or external metrics
+source](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#support-for-metrics-apis)
+in a kubernetes cluster. This is a problem because the different custom
+metric providers which have been developed since this feature was added
+provide different metrics for different domains. For instance the
+[prometheus-metrics-adapter](https://github.com/kubernetes-sigs/prometheus-adapter)
+exposes metrics stored in Prometheus and [KEDA](https://github.com/kedacore/keda) uses external metrics for event driven based autoscaling. Similarly there are metrics adapters
+for different metric sources like
+[stackdriver](https://github.com/GoogleCloudPlatform/k8s-stackdriver/tree/master/custom-metrics-stackdriver-adapter),
+[AWS Cloudwatch](https://github.com/awslabs/k8s-cloudwatch-adapter). This
+poses a problem because cluster administrators have to choose only a single
+metric adapter to support in their clusters. The proposed solution is to
+create a router which acts as the single entry point for all metrics which
+then routes the metric request to the correct metrics provider based on
+existing information already exposed in most metrics adapters. The router can
+leverage existing projects like the
+[custom-metrics-apiserver](https://github.com/kubernetes-sigs/custom-metrics-apiserver)
+and the [metrics client](https://github.com/kubernetes/metrics).
+
+### Goals
+
+* Allows for multiple metric sources in a cluster
+* Existing metrics sources should be compatible with solution
+* Leverage existing functionality in metric sources for discovery and routing.
+
+### Non-Goals
+
+* Configuring metric sources at a fine-grain level
+
+## Proposal
+
+### Background
+
+#### custom-metrics-apiserver
+This [project](https://github.com/kubernetes-sigs/custom-metrics-apiserver)
+provides helper classes to build custom/external metric providers. Most
+custom/external metric providers use this framework. It is important to note
+here that the metrics server is an aggregated API server which provides a lot
+of the functionality which is present in the kube-apiserver like API
+discovery and RBAC authentication/authorization.
+
+#### Metrics Client
+The [API and client](https://github.com/kubernetes/metrics) for resource,
+custom and external metrics are used by the HPA controller to query for
+metrics from any registered custom/external metric providers. Technically the
+client is instantiated with the kube-apiserver as the target for the requests
+which in turn forwards it to the metrics server based on the API group i.e.
+`metrics.k8s.io`, `custom.metrics.k8s.io` or `external.metrics.k8s.io`.
+
+#### API Discovery
+API servers expose endpoints which can be used by a client to find the
+correct resource version and discover information about which API resources
```

Comment:
The router would rely on the official [metrics-client](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/metrics) and I assume that the client uses the discovery API to find the preferred version and then query that. I might be wrong. Is there any use case you can think of which would require an override of the default behavior.

Reply:
> Is there any use case you can think of which would require an override of the default behavior.

No, I think you're right. Thanks for the clarification 👍

Reply:
> The router would rely on the official [metrics-client](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/metrics) and I assume that the client uses the discovery API to find the preferred version and then query that

Sorry, I remember now 😄 : I'm not sure it is possible to use the metrics-client to discover the list of metrics.
For example in [your proof concept](https://github.com/arjunrn/custom-metrics-router/blob/ade1ceb84ff9a78dc5948efd8ec9a8d325266aae/pkg/metricsclient/metricsclient.go#L103) you use the discovery client:

```golang
c.discoveryClient.ServerResourcesForGroupVersion(customMetricsAPI.SchemeGroupVersion.String())
```

In the snippet above `customMetricsAPI.SchemeGroupVersion.String()` would be statically resolved according to the imported version of `customMetricsAPI`. Which might not be the same than the one selected by the metrics client.

But I don't think it's a problem since we can reuse the `AvailableAPIsGetter` from the metrics client, for example if we keep `cmClient.NewAvailableAPIsGetter(discoveryClient)` in the [Client structure](https://github.com/arjunrn/custom-metrics-router/blob/master/pkg/metricsclient/metricsclient.go#L36):

```golang
version, err := c.customMetricsAvailableAPIsGetter.PreferredVersion()
if err != nil {
	return nil, err
}
resources, err := c.discoveryClient.ServerResourcesForGroupVersion(version.String())
if err != nil {
	return nil, fmt.Errorf("failed to get resource for %s: %v", customMetricsAPI.SchemeGroupVersion, err)
}
```

But this is an implementation detail, and you can still ignore my initial comment in the context of the KEP.